framework/data/BaseDataProvider.php

```diff
@@ -44,6 +50,20 @@
     private $_keys;
     private $_models;
     private $_totalCount;
+    
+    /**
+     * @inheritdoc
+     */
+    public function init()
+    {
+        parent::init();
+        if (!isset($this->id)) {
+            if (self::$counter > 0) {
+                $this->id = 'dp-' . self::$counter;
```

Comment:
@SilverFire Since I have to open a new issue for the refactoring proposed by @dynasource, would you be okay with merging this and discussing custom generation of ids there?

It is also a common thing in the framework and affects implementation of the trait as well. If we add something to the public API now it'll be a BC break if we change it.